tools/dev-doctor/check.go

```diff
@@ -0,0 +1,41 @@
+// Copyright 2020 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+// A checkResult is the result of a check.
+type checkResult int
+
+const (
+	checkSkipped checkResult = -1 // The check was skipped.
+	checkOK      checkResult = 0  // The check completed and did not find any problems.
+	checkWarning checkResult = 1  // The check completed and found something that might indicate a problem.
+	checkError   checkResult = 2  // The check completed and found a definite problem.
+	checkFailed  checkResult = 3  // The check could not be completed.
+)
+
+// A check is an individual check.
+type check interface {
+	Name() string               // Name returns the check's name.
+	Run() (checkResult, string) // Run runs the check.
```

Comment:
Nit: generally, the last return value is an `error`. Here I notice the `string` represents the error string anyway, which is then printed when the tool finishes all its checks. While this works, it's not a "typical" Go. Alternatively, this code can return `error` (and it won't need to retrieve & return the string representation on each instance), and just simply format the final output once.

Reply:
The execution flow in this program with respect to errors is a little different to normal. Normally, when you encounter an error you abandon what you're doing and return to the caller. When running all these checks, you want to keep going and move on to the next check, even if an individual check fails.

There's also a difference here that the last value returned is not an error, it is a message to be shown to the user.

In summary, a check does not return an error in the usual Go sense of an error. It always returns a result and a message, and in the worst case that result can be "failed" and the message explains why.

Hope this makes sense.